xla/service/gpu/runtime/copy_thunk.cc

```diff
@@ -45,41 +43,112 @@ absl::Status DeviceToDeviceCopyThunk::ExecuteOnStream(
-    ThunkInfo thunk_info, const BufferAllocation::Slice& source_buffer,
-    const BufferAllocation::Slice& destination_buffer, uint64_t mem_size)
-    : DeviceToDeviceCopyThunk(thunk_info, source_buffer, destination_buffer,
-                              mem_size) {}
+//===----------------------------------------------------------------------===//
+// CopyAsyncEvents
+//===----------------------------------------------------------------------===//
 
-absl::Status DeviceToHostCopyThunk::ExecuteOnStream(
-    const ExecuteParams& params) {
-  se::DeviceMemoryBase destination_data =
-      params.buffer_allocations->GetDeviceAddress(destination());
-  se::DeviceMemoryBase source_data =
-      params.buffer_allocations->GetDeviceAddress(source());
-  void* cpu_dst = destination_data.opaque();
-  VLOG(3) << "Memcpy D2H for memory offload from " << source_data.opaque()
-          << " to " << cpu_dst;
-  return params.stream->Memcpy(cpu_dst, source_data, size_bytes());
+// Emplace() will insert {key, event} pair into the hash map,
+// and return the event in order to do RecordEvent() for async memcpy.
+absl::Status CopyAsyncEvents::Emplace(se::StreamExecutor *executor,
+                                      const HloInstruction *instr,
+                                      se::Event &&event) {
+  Key key = {executor, instr};
+
+  absl::MutexLock lock(&mutex_);
+  if (auto [it, inserted] = events_.try_emplace(key, std::move(event));
+      inserted) {
+    return absl::OkStatus();
+  }
+  return absl::InternalError("Async copy event already exists!");
 }
 
-HostToDeviceCopyThunk::HostToDeviceCopyThunk(
-    ThunkInfo thunk_info, const BufferAllocation::Slice& source_buffer,
-    const BufferAllocation::Slice& destination_buffer, uint64_t mem_size)
+// Retrieve a completion event started by copy-start instruction
+// `instr`, and remove the event from the collection.
+absl::StatusOr<se::Event>
+CopyAsyncEvents::Extract(se::StreamExecutor *executor,
+                         const HloInstruction *instr) {
+
+  Key key = {executor, instr};
+  absl::MutexLock lock(&mutex_);
+  if (auto event = events_.extract(key)) {
+    return std::move(event.mapped());
+  }
+  return absl::InternalError("Async copy event was not found!");
+}
+
+//===----------------------------------------------------------------------===//
+// DeviceHostCopyThunk
+//===----------------------------------------------------------------------===//
+DeviceHostCopyThunk::DeviceHostCopyThunk(
+    ThunkInfo thunk_info, const BufferAllocation::Slice &source_buffer,
+    const BufferAllocation::Slice &destination_buffer, uint64_t mem_size,
+    std::shared_ptr<CopyAsyncEvents> async_events, const HloInstruction *instr,
+    bool device_to_host)
     : DeviceToDeviceCopyThunk(thunk_info, source_buffer, destination_buffer,
-                              mem_size) {}
+                              mem_size),
+      stream_id_(thunk_info.execution_stream_id),
```

Comment:
so when you call WithProfileAnnotation when creating the thunk in ir emitter, it will already set the execution stream id in the base Thunk class, you dont need to save another copy here. You can get the stream id by calling Thunk::execution_stream_id()

Reply:
Will remove.